nova/virt/vmwareapi/driver.py

```diff
@@ -693,11 +695,288 @@ def detach_interface(self, context, instance, vif):
+        :param network_info: instance network information
+        :param disk_info: instance disk information
+        :param migrate_data: a LiveMigrateData object
+        :returns: migrate_data modified by the driver
+        """
+        result = self._vmops.place_vm(context, instance)
+        if result.recommendations:
+            rs = sorted([r for r in result.recommendations
+                         if r.reason == "xvmotionPlacement" and
+                            r.action],
+                        key=lambda r: r.rating)
+            if rs:
+                relocate_spec = rs[0].action[0].relocateSpec
+
+        if not relocate_spec:
+            raise exception.MigrationError(
+                reason="Could not place VM")
+
+        if not relocate_spec.host:
+            raise exception.MigrationError(
+                reason="No host with enough resources")
+
+        if not relocate_spec.datastore:
+            raise exception.MigrationError(
+                reason="No datastore with enough resources")
+
+        # relocate_defaults are serialised/deserialised on put/get
+        defaults = migrate_data.relocate_defaults
+        spec = vim_util.serialize_object(relocate_spec)
+        defaults["relocate_spec"] = spec
+        # Writing the values back
+        migrate_data.relocate_defaults = defaults
+
+        return migrate_data
+
+    def live_migration(self, context, instance, dest,
+                       post_method, recover_method, block_migration=False,
+                       migrate_data=None):
+        """Live migration of an instance to another host.
+
+        :param context: security context
+        :param instance:
+            nova.db.sqlalchemy.models.Instance object
+            instance object that is migrated.
+        :param dest: destination host
+        :param post_method:
+            post operation method.
+            expected nova.compute.manager._post_live_migration.
+        :param recover_method:
+            recovery method when any exception occurs.
+            expected nova.compute.manager._rollback_live_migration.
+        :param block_migration: if true, migrate VM disk.
+        :param migrate_data: a LiveMigrateData object
+
+        """
+        dest_session = self._create_dest_session(migrate_data)
+        try:
+            disk_mappings = self._get_disk_mappings(context, instance)
+            self._set_vif_infos(instance, migrate_data, dest_session)
+            self._vmops.live_migration(instance, migrate_data, disk_mappings)
+            LOG.info("Migration operation has completed", instance=instance)
+            post_method(context, instance, dest, block_migration, migrate_data)
+        except exception.InstanceNotFound:
+            # Try to get the instance on the remote session
+            try:
+                vm_util.get_vm_ref.__wrapped__(dest_session, instance)
+                # TODO(fwiesel): get the migration task
```

Comment:
In case we restart the migration, and the migration is still ongoing, then I'd rather want to wait until the migration finishes, before declaring it complete in openstack.

Reply:
I'd prefer this explanation in the code/comment, as we might only pick it up much later and then maybe someone else will or you won't remember what exactly you meant. :)